app/uk/gov/hmrc/eusubsidycompliancefrontend/cache/UndertakingCache.scala

```diff
@@ -0,0 +1,111 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.eusubsidycompliancefrontend.cache
+
+import org.mongodb.scala.MongoCollection
+import org.mongodb.scala.model.{Filters, IndexOptions, Indexes, Updates}
+import play.api.libs.json.{Reads, Writes}
+import uk.gov.hmrc.eusubsidycompliancefrontend.cache.UndertakingCache.DefaultCacheTtl
+import uk.gov.hmrc.eusubsidycompliancefrontend.models.types.{EORI, UndertakingRef}
+import uk.gov.hmrc.mongo.cache.{CacheIdType, CacheItem, DataKey, MongoCacheRepository}
+import uk.gov.hmrc.mongo.{CurrentTimestampSupport, MongoComponent}
+
+import javax.inject.{Inject, Singleton}
+import scala.concurrent.duration.{DurationInt, FiniteDuration}
+import scala.concurrent.{ExecutionContext, Future}
+import scala.language.postfixOps
+import scala.reflect.ClassTag
+
+object EoriIdType extends CacheIdType[EORI] {
+  override def run: EORI => EORI = identity
+}
+
+@Singleton
+class UndertakingCache @Inject() (
+  mongoComponent: MongoComponent
+)(implicit ec: ExecutionContext) extends MongoCacheRepository[EORI](
+    mongoComponent = mongoComponent,
+    collectionName = "undertakingCache",
+    ttl = DefaultCacheTtl,
+    timestampSupport = new CurrentTimestampSupport,
+    cacheIdType = EoriIdType
+  ) {
+
+  private val UndertakingReference = "data.Undertaking.reference"
+  private val UndertakingSubsidiesIdentifier = "data.UndertakingSubsidies.undertakingIdentifier"
+
+  // Ensure additional indexes for undertaking and undertaking subsidies deletion are present.
+  private lazy val indexedCollection: Future[MongoCollection[CacheItem]] =
+    for {
+      _ <- collection.createIndex(
+        Indexes.ascending(UndertakingReference),
+        IndexOptions()
+          .background(false)
+          .name("undertakingReference")
+          .sparse(false)
+          .unique(false)
+      ).headOption()
+      _ <- collection.createIndex(
+        Indexes.ascending(UndertakingSubsidiesIdentifier),
+        IndexOptions()
+          .background(false)
+          .name("undertakingSubsidiesIdentifier")
+          .sparse(false)
+          .unique(false)
+      ).headOption()
+    } yield collection
+
+  def get[A : ClassTag](eori: EORI)(implicit reads: Reads[A]): Future[Option[A]] =
+    indexedCollection.flatMap { _ =>
+      super
+        .get[A](eori)(dataKeyForType[A])
+      }
+
+  def put[A](eori: EORI, in: A)(implicit writes: Writes[A]): Future[A] =
+    indexedCollection.flatMap { _ =>
+      super
+        .put[A](eori)(DataKey(in.getClass.getSimpleName), in)
+        .map(_ => in)
+    }
+
+  def deleteUndertaking(ref: UndertakingRef): Future[Unit] =
```

Comment:
Instead of delete can it be renamed to update? Delete seems a bit misleading

Reply:
It is a cache delete. Mongo requires an update in order to delete an object within a document. This is consistent with the implementation of the HMRC MongoCacheRepository.